### PR TITLE
WebTools (WIP): add ThrustExpo tool

### PR DESCRIPTION
**Pull request by yuri-rage**
Uses thrust test stand data to autofit thrust linearization and arrive at a proper MOT_THST_EXPO value.

Please double check the arithmetic, as it does appear correct, but I'm finding that the example data I use (which is sparse from the forum) results in much lower values than the ones the spreadsheet uses. I think this is explained by constraining the dataset to SPIN_MIN through SPIN_MAX and using a more precise approach overall. Lines 208 and 298 of `ThrustExpo.js` are the primary drivers for the curve fit.

Minor change to `ParameterMetaData.js` allows for optionally constraining numeric input elements to the high/low range defined in the JSON files by adding a "constrain" class to the element attributes (non-breaking for existing pages).

Tested on Chromium and Firefox based browsers with identical results.

### CODE REVIEW

**Comment by IamPete1** [comment]
The Github logo sticks to the far right of the page, looks odd on big monitors:
image

I the other headers there is a max width of 1200px. 

Like wise I think you can increase the max width of the content to 1200px to match the other tools.

**Review by IamPete1 — COMMENTED** [review]
My preference would be that the plots all all there when you first load the page with no data. Adding stuff at runtime makes everything move about.

I'm going to play about with the fit and will push a branch.

File: ThrustExpo/index.html [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+                                disabled
+                            />
+                        </div>
+                    </div>
+                    <div class="param-column">
+                        <div
+                            class="param-row"
+                            title="Total mass in kg. 0 to disable hover thrust estimate."
+                        >
+                            <label>Mass (with battery)</label>
```
[the remark this answers is not in the document]
Comment (yuri-rage) [reply ThrustExpo/index.html:194]: Maybe All Up Weight (AUW) - the technically correct term that I butchered in favor of metric units 😄

File: ThrustExpo/ThrustExpo.js [hunk 2]
```diff
@@ -0,0 +1,890 @@
...
+        xaxis: {
+            title: { text: "Throttle (%)" },
+            type: "linear",
+            zeroline: false,
+            showline: true,
+            mirror: true,
+            range: [0, 100],
+        },
+        yaxis: {
+            title: { text: "Thrust (kgf)" },
```
[the remark this answers is not in the document]
Comment (yuri-rage) [reply ThrustExpo/ThrustExpo.js:543]: Too easy.

**Comment by yuri-rage** [comment]
@IamPete1, point taken about max-width. I was attempting a little more responsive page design, and I do use a very wide monitor but hadn't maximized the page to see how wonky it gets.

I did like keeping the plots to a more square aspect, as I think it makes the linear fit easier to visualize than stretching it across the screen.

As for the plots appearing at runtime - this is the only tool where a big table of data has to be user-entered. I didn't want to put that below a bunch of empty plots, as it would be obscured. And I didn't want the plots to go below a potentially lengthy table that grows (to a limit) as data is entered. And I like displaying more than a few lines of that table so you can eyeball it for errors or value trends. I'll have a go at rearranging the page a bit and see how we can compromise a little.

**Comment by rmackay9** [comment]
It's great to see this.   I think replacing the spreadsheet would make it much easier for people to use.

Perhaps for now we just want a one-to-one replacement but I suspect we could automatically calculate the expo that minimises the error between the actual and predicted.

**Comment by rmackay9** [comment]
By the way, it looks like there's a merge commit included that we probably want to get rid of before merging

**Comment by yuri-rage** [comment]
> Perhaps for now we just want a one-to-one replacement but I suspect we could automatically calculate the expo that minimises the error between the actual and predicted.

It's already quite a bit more than a 1:1 replacement, and it does calculate an optimized value (though that "optimization" is a little naive). However, the datasets I can find to test against show very promising results, and I suspect many of us are using values that are a bit too high due to the lack of granularity and method used in the spreadsheet.

**Review by IamPete1 — APPROVED** [review]
Thanks for all your work on this!

**Pull request merged**